File: app/src/main/java/arun/com/chromer/browsing/article/BaseArticleActivity.java
/*
 * Chromer
 * Copyright (C) 2017 Arunkumar
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package arun.com.chromer.browsing.article;

import android.annotation.TargetApi;
import android.app.ActivityManager;
import android.graphics.Bitmap;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatDelegate;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.ProgressBar;

import com.bumptech.glide.request.target.SimpleTarget;
import com.bumptech.glide.request.transition.Transition;

import javax.inject.Inject;

import arun.com.chromer.R;
import arun.com.chromer.browsing.article.view.ElasticDragDismissFrameLayout;
import arun.com.chromer.data.webarticle.WebArticleRepository;
import arun.com.chromer.data.webarticle.model.WebArticle;
import arun.com.chromer.shared.base.activity.BaseActivity;
import arun.com.chromer.util.glide.GlideApp;
import rx.Subscription;
import rx.subscriptions.CompositeSubscription;

public abstract class BaseArticleActivity extends BaseActivity {
    private static final String TAG = BaseArticleActivity.class.getSimpleName();
    private static final boolean DEBUG = false;

    private static final int MIN_NUM_ELEMENTS = 1;

    private String url;

    private int primaryColor;
    private int accentColor;


    private final CompositeSubscription subscriptions = new CompositeSubscription();

    @Inject
    WebArticleRepository webArticleRepository;

    private RecyclerView recyclerView;
    private ArticleAdapter adapter;

    private final View.OnClickListener sideClickListener = view -> finish();
    private ProgressBar progressBar;
    private Toolbar toolbar;
    private ArticleScrollListener articleScrollListener;

    @Override
    protected int getLayoutRes() {
        return R.layout.article_activity_article;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        url = getIntent().getDataString();

        if (DEBUG) {
            Log.v(TAG, "loading article: " + url);
        }
        setDayNightTheme();
        readCustomizations();

        toolbar = findViewById(R.id.article_toolbar);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setHomeAsUpIndicator(R.drawable.article_ic_close);
            getSupportActionBar().setTitle(null);
        }

        final View statusBar = findViewById(R.id.article_status_bar);
        recyclerView = findViewById(R.id.article_recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        articleScrollListener = new ArticleScrollListener(toolbar, statusBar, primaryColor);
        recyclerView.addOnScrollListener(articleScrollListener);

        progressBar = findViewById(R.id.article_loading);

        findViewById(R.id.article_transparent_side_1).setOnClickListener(sideClickListener);
        findViewById(R.id.article_transparent_side_2).setOnClickListener(sideClickListener);

        final ElasticDragDismissFrameLayout dragDismissLayout = findViewById(R.id.article_drag_dismiss_layout);
        dragDismissLayout.addListener(new ElasticDragDismissFrameLayout.ElasticDragDismissCallback() {
            @Override
            public void onDragDismissed() {
                super.onDragDismissed();
                finish();
            }
        });

        final Subscription subscription = webArticleRepository
                .getWebArticle(url)
                .doOnError(this::onArticleLoadingFailed)
                .subscribe(webArticle -> {
                    if (webArticle == null) {
                        onArticleLoadingFailed(null);
                    } else {
                        onArticleLoaded(webArticle);
                    }
                });
        subscriptions.add(subscription);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        subscriptions.clear();
    }

    private void setPrimaryColor(@ColorInt int primaryColor) {
        this.primaryColor = primaryColor;
        ArticleUtil.changeRecyclerOverscrollColors(recyclerView, primaryColor);
        ArticleUtil.changeProgressBarColors(progressBar, primaryColor);
        articleScrollListener.setPrimaryColor(primaryColor);
    }

    private void readCustomizations() {
        primaryColor = getIntent().getIntExtra(ArticleIntent.EXTRA_TOOLBAR_COLOR,
                ContextCompat.getColor(this, R.color.article_colorPrimary));
        accentColor = getIntent().getIntExtra(ArticleIntent.EXTRA_ACCENT_COLOR,
                ContextCompat.getColor(this, R.color.article_colorAccent));
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
        }
        return super.onOptionsItemSelected(item);
    }

    private void setDayNightTheme() {
        int theme = getIntent().getIntExtra(ArticleIntent.EXTRA_THEME, ArticleIntent.THEME_AUTO);
        if (theme == ArticleIntent.THEME_LIGHT) {
            getDelegate().setLocalNightMode(AppCompatDelegate.MODE_NIGHT_NO);
        } else if (theme == ArticleIntent.THEME_DARK) {
            getDelegate().setLocalNightMode(AppCompatDelegate.MODE_NIGHT_YES);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                getWindow().setNavigationBarColor(getResources().getColor(R.color.article_windowBackground));
            }
        } else {
            getDelegate().setLocalNightMode(AppCompatDelegate.MODE_NIGHT_AUTO);
        }
    }

    protected void onArticleLoadingFailed(@Nullable Throwable throwable) {
        progressBar.setVisibility(View.GONE);
    }

    protected void onArticleLoaded(@NonNull WebArticle webArticle) {
        if (webArticle.elements != null && webArticle.elements.size() >= MIN_NUM_ELEMENTS) {
            renderArticle(webArticle);
            applyTaskDescriptionAndColors(webArticle);
        } else {
            onArticleLoadingFailed(null);
        }
    }

    private void applyTaskDescriptionAndColors(@NonNull final WebArticle webArticle) {
        final String title = webArticle.safeLabel();
        final String faviconUrl = webArticle.faviconUrl;
        final int color = webArticle.themeColor() == -1 ? primaryColor : webArticle.themeColor();
        setPrimaryColor(color);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setTaskDescription(new ActivityManager.TaskDescription(title, null, color));
            GlideApp.with(this)
                    .asBitmap()
                    .load(faviconUrl)
                    .into(new SimpleTarget<Bitmap>() {
                        @TargetApi(Build.VERSION_CODES.LOLLIPOP)
                        @Override
                        public void onResourceReady(Bitmap resource, Transition<? super Bitmap> transition) {
                            setTaskDescription(new ActivityManager.TaskDescription(title, resource, color));
                        }
                    });
        }
    }

    private void renderArticle(@NonNull WebArticle webArticle) {
        adapter = new ArticleAdapter(webArticle, accentColor);
        recyclerView.setAdapter(adapter);
        progressBar.setVisibility(View.GONE);
        adapter.addElements(webArticle.elements);
    }
}

File: app/src/main/java/arun/com/chromer/browsing/article/BaseArticleActivity.kt
/*
 * Chromer
 * Copyright (C) 2017 Arunkumar
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package arun.com.chromer.browsing.article

import android.arch.lifecycle.ViewModelProviders
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.support.annotation.CallSuper
import android.support.annotation.ColorInt
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatDelegate
import android.support.v7.widget.LinearLayoutManager
import arun.com.chromer.R
import arun.com.chromer.browsing.BrowsingActivity
import arun.com.chromer.browsing.article.adapter.ArticleAdapter
import arun.com.chromer.browsing.article.util.ArticleScrollListener
import arun.com.chromer.browsing.article.util.ArticleUtil.changeProgressBarColors
import arun.com.chromer.browsing.article.util.ArticleUtil.changeRecyclerOverscrollColors
import arun.com.chromer.browsing.article.view.ElasticDragDismissFrameLayout
import arun.com.chromer.browsing.menu.MenuDelegate
import arun.com.chromer.data.Result
import arun.com.chromer.data.webarticle.model.WebArticle
import arun.com.chromer.data.website.model.Website
import arun.com.chromer.extenstions.gone
import arun.com.chromer.extenstions.setMenuBackgroundColor
import arun.com.chromer.settings.Preferences
import arun.com.chromer.settings.Preferences.*
import arun.com.chromer.shared.Constants
import arun.com.chromer.shared.Constants.EXTRA_KEY_TOOLBAR_COLOR
import arun.com.chromer.tabs.DefaultTabsManager
import arun.com.chromer.util.Utils
import arun.com.chromer.util.glide.GlideApp
import it.sephiroth.android.library.bottomnavigation.BottomNavigation
import kotlinx.android.synthetic.main.activity_article_mode.*
import javax.inject.Inject

abstract class BaseArticleActivity : BrowsingActivity() {
    private lateinit var browsingArticleViewModel: BrowsingArticleViewModel

    private var url: String? = null

    private var primaryColor: Int = 0
    private var accentColor: Int = 0

    private lateinit var articleAdapter: ArticleAdapter

    private var articleScrollListener: ArticleScrollListener? = null

    @Inject
    lateinit var tabsManager: DefaultTabsManager
    @Inject
    lateinit var menuDelegate: MenuDelegate

    override fun getLayoutRes(): Int {
        return R.layout.activity_article_mode
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        readCustomizations()
        url = intent.dataString

        setupToolbar()
        setupCloseListeners()
        setupBottombar()
        setupTheme()

        articleScrollListener = ArticleScrollListener(toolbar, statusBar, primaryColor)
        recyclerView.addOnScrollListener(articleScrollListener)
    }

    private fun setupCloseListeners() {
        transparentSide1.setOnClickListener { finish() }
        transparentSide2.setOnClickListener { finish() }
        dragDismissLayout.addListener(
                object : ElasticDragDismissFrameLayout.ElasticDragDismissCallback() {
                    override fun onDragDismissed() {
                        finish()
                    }
                })
    }


    private fun setupBottombar() {
        bottomNavigation.setMenuBackgroundColor(ContextCompat.getColor(this, R.color.article_windowBackground))

        if (preferences.bottomBar()) {
            bottomNavigation.apply {
                setSelectedIndex(-1, false)
                setOnMenuItemClickListener(object : BottomNavigation.OnMenuItemSelectionListener {
                    override fun onMenuItemSelect(itemId: Int, position: Int, fromUser: Boolean) {
                        menuDelegate.handleItemSelected(itemId)
                        post { setSelectedIndex(-1, false) }
                    }

                    override fun onMenuItemReselect(itemId: Int, position: Int, fromUser: Boolean) {
                    }
                })
            }
        } else {
            bottomNavigation.gone()
        }
    }

    private fun setupTheme() {
        val theme = preferences.articleTheme()
        if (theme == Preferences.THEME_BLACK) {
            coordinatorLayout.background = ColorDrawable(Color.BLACK)
            bottomNavigation.setMenuBackgroundColor(Color.BLACK)
        }
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        browsingArticleViewModel = ViewModelProviders.of(this, viewModelFactory).get(BrowsingArticleViewModel::class.java)
        subs.add(browsingArticleViewModel.loadWebSiteDetails(intent.dataString)
                .subscribe { result ->
                    when (result) {
                        is Result.Success -> {
                            val webArticle: WebArticle? = result.data
                            if (webArticle == null) {
                                onArticleLoadingFailed(null)
                            } else {
                                onArticleLoaded(webArticle)
                            }
                        }
                        is Result.Failure -> onArticleLoadingFailed(result.throwable)
                    }
                })
    }


    private fun setupToolbar() {
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setHomeAsUpIndicator(R.drawable.article_ic_close)
        supportActionBar?.title = null
    }

    override fun onWebsiteLoaded(website: Website) {
        if (website.themeColor() != Constants.NO_COLOR) {
            setPrimaryColor(website.themeColor())
        }
    }

    private fun setPrimaryColor(@ColorInt primaryColor: Int) {
        if (preferences.dynamiceToolbarEnabledAndWebEnabled()) {
            this.primaryColor = primaryColor
            changeRecyclerOverscrollColors(recyclerView, primaryColor)
            changeProgressBarColors(progressBar, primaryColor)
            articleScrollListener?.setPrimaryColor(primaryColor)
        }
    }

    private fun readCustomizations() {
        val appPrimaryColor = preferences.toolbarColor()
        primaryColor = intent.getIntExtra(EXTRA_KEY_TOOLBAR_COLOR, appPrimaryColor)
        accentColor = ContextCompat.getColor(this, R.color.accent)

        if (preferences.dynamiceToolbarEnabledAndWebEnabled()) {
            accentColor = primaryColor
        }

        val theme = preferences.articleTheme()
        when (theme) {
            THEME_LIGHT -> delegate.setLocalNightMode(AppCompatDelegate.MODE_NIGHT_NO)
            THEME_DARK, THEME_BLACK -> delegate.setLocalNightMode(AppCompatDelegate.MODE_NIGHT_YES)
            else -> delegate.setLocalNightMode(AppCompatDelegate.MODE_NIGHT_AUTO)
        }
    }

    private fun renderArticle(webArticle: WebArticle) {
        articleAdapter = ArticleAdapter(
                webArticle,
                accentColor,
                GlideApp.with(this)
        ).apply {
            setElements(webArticle.elements)
            subs.add(keywordsClicks().map { Utils.getSearchUrl(it) }.subscribe {
                tabsManager.openUrl(this@BaseArticleActivity, Website(it))
            })
        }

        recyclerView.apply {
            layoutManager = LinearLayoutManager(this@BaseArticleActivity)
            adapter = articleAdapter
        }
        hideLoading()
    }

    @CallSuper
    protected open fun onArticleLoadingFailed(throwable: Throwable?) {
        hideLoading()
    }

    protected open fun onArticleLoaded(webArticle: WebArticle) {
        if (webArticle.elements != null && webArticle.elements.size >= MIN_NUM_ELEMENTS) {
            renderArticle(webArticle)
        } else {
            onArticleLoadingFailed(null)
        }
    }

    override fun onToolbarColorSet(websiteThemeColor: Int) {
    }

    private fun hideLoading() {
        progressBar.gone()
    }

    companion object {
        private const val MIN_NUM_ELEMENTS = 1
    }
}
